### PR TITLE
Embed all header files (as zip) in the PCH.

### PR DESCRIPTION
**Pull request by vgvassilev**
This should reduce the amount of patches we have in clang making the PCH
relocatable.

### CODE REVIEW

**Comment by dpiparo** [comment]
Hi @vgvassilev, 

nice and clean! I have two questions:
1) why do you say "should"?
2) does this imply that the files in the pch can also be absent from the disk?

**Comment by vgvassilev** [comment]
I say 'should' because that's work in progress. Once I get a clean slate I will start reverting some patches in clang. Yes, it should be possible to get rid of the files on disk.

File: core/dictgen/src/rootcling_impl.cxx [hunk 1]
```diff
@@ -4221,6 +4221,9 @@ int RootClingMain(int argc,
    clingArgs.push_back("-Xclang");
    clingArgs.push_back("-main-file-name");
    clingArgs.push_back("-Xclang");
+   // FIXME: Enable this once we switch cling to use the driver.
+   clingArgs.push_back("-fmodules-embed-all-files");
+   clingArgs.push_back("-Xclang");
    clingArgs.push_back((dictname + ".h").c_str());
```
Comment (Axel-Naumann) [inline core/dictgen/src/rootcling_impl.cxx]: This breaks the parameters. We want `--cc1` to see `-main-file-name dictname.h -fmodules-embed-all-files`; this change makes that `-main-file-name -fmodules-embed-all-files dictname.h` i.e. the main file name is now broken.

File: interpreter/llvm/src/tools/clang/lib/Serialization/ASTReader.cpp [hunk 2]
```diff
@@ -1244,37 +1244,8 @@ bool ASTReader::ReadSourceManagerBlock(ModuleFile &F) {
 /// \brief Attempt to resolve the location based on PP header search.
 /// Find the first match with the longest trailing part.
 static StringRef
-resolveFileThroughHeaderSearch(Preprocessor& PP, StringRef Filename,
-                               llvm::StringMap<std::string>& HSStemMap) {
-  // Locate /a/b/c/d.h given the -I paths, by removing leading subdirectories
-  // until a file is found.
+resolveFileThroughHeaderSearch(Preprocessor& PP, StringRef Filename) {
```
Comment (Axel-Naumann) [inline interpreter/llvm/src/tools/clang/lib/Serialization/ASTReader.cpp]: I bet also 6545d8c211 e893dfcd36 7e31f48303 8240dcfe4b (hashes from root.git) can be reverted. I'm not sure whether reverting them or not reverting them will result in a more broken state, once you move all zipped headers into the module...

File: core/dictgen/src/rootcling_impl.cxx (hunk 1 above)
Comment (vgvassilev) [reply core/dictgen/src/rootcling_impl.cxx]: Sure, fixed.

File: interpreter/llvm/src/tools/clang/lib/Serialization/ASTReader.cpp (hunk 2 above)
Comment (vgvassilev) [reply interpreter/llvm/src/tools/clang/lib/Serialization/ASTReader.cpp]: That's work in progress, I am intending to do that and then we will see.

Comment (vgvassilev) [reply interpreter/llvm/src/tools/clang/lib/Serialization/ASTReader.cpp]: Btw, it looks like 6545d8c and e893dfc are not active anymore.

**Pull request merged**